pilot/pkg/model/push_context.go

```diff
@@ -945,7 +945,18 @@ func (ps *PushContext) VirtualServicesForGateway(proxyNamespace, gateway string)
 		len(ps.virtualServiceIndex.publicByGateway[gateway]))
 	res = append(res, ps.virtualServiceIndex.privateByNamespaceAndGateway[name]...)
 	res = append(res, ps.virtualServiceIndex.exportedToNamespaceByGateway[name]...)
-	res = append(res, ps.virtualServiceIndex.publicByGateway[gateway]...)
+	// Favor same-namespace Gateway routes, to give the "consumer override" preference.
+	// We do 2 iterations here to avoid extra allocations.
+	for _, vs := range ps.virtualServiceIndex.publicByGateway[gateway] {
+		if UseGatewaySemantics(vs) && vs.Namespace == proxyNamespace {
+			res = append(res, vs)
+		}
+	}
+	for _, vs := range ps.virtualServiceIndex.publicByGateway[gateway] {
+		if !(UseGatewaySemantics(vs) && vs.Namespace == proxyNamespace) {
+			res = append(res, vs)
+		}
+	}
```

Comment:
I am sure this will break if a vs not in same namespace with gateway and the proxy

Reply:
Not as the pr described, it does not respect dest service namespace

Reply:
I am not sure what destination service namespace has to do with anything. The precedence is to use your own namespace routes before other namespaces. This change is pretty extensively tested. If there is an issue can you give an example yaml that behaves incorrectly?